### ISSUE TITLE
Should `updateChatLastMessage` positions be considered for?

### ISSUE THREAD

**Issue by vincentneo** [tdlib/td#2545]
I have a case where through some basic logging, seems to suggests that a `updateChatLastMessage` that happened after a `updateChatPosition`, reported `positions` for a chat that was outdated.

In particular, in one observation, `updateChatPosition` reported a position order of 0, while `updateChatLastMessage` reported a position order of 7258204463465589541. 

The folder in which the user is at (or I should say the folder of interest in this case), is one that works as a folder that includes everything and excludes read messages (unread folder). The particular chat that I mentioned, reports an unread messages value of 0, which suggests the `updateChatPosition`'s value to be the correct one.

Which begs the question, should `updateChatLastMessage`'s positions be considered for? How safe is it to disregard it? ~~(A quick test suggests incorrect ordering, so that kinda answered those questions.)~~ (My quick test earlier seems flawed, leaving the questions open)

Context: I am investigating a case where the above mentioned unread folder, sometimes still contains chats that have unread count of 0. It is known to my best knowledge, that issue only happens when chat is fully read on another device, not when read on the TDLib-based client.

Implementation covers position updates from `updateChatLastMessage`, `updateChatPosition` and `updateChatDraftMessage`, as recommended in [td_api.tl, line 5538](https://github.com/tdlib/td/blob/f817d79b7d98670a3a824ec48dfb6fd9c4b7e9ef/td/generate/scheme/td_api.tl#L5538C54-L5538C54)

Note: final TDLib 1.8.14 with relevant patches was used when this was observed, not 1.8.15 released recently.

PS: I will not be able to provide TDLib logs for this case, as issue happened on client device, and I failed to ever replicate this personally.

**Comment by levlam** [comment]
Updates about chat positions must not be ignored and must be applied strictly in the order they are received. In this case all chat positions should be correct. Check that you correctly apply `updateChatPosition` with `position.order == 0` by removing the chat from the list.

**Comment by vincentneo** [comment]
> In this case all chat positions should be correct. 

As I've said in my original comment, the position delivered by updateChatLastMessage was incorrect, which came after the updateChatPosition.

updateChatPosition came with a position order of 0, which means I should remove, but updateChatLastMessage came with a position order number.

By the way, it should be noted that updateChatLastMessage came with 3 positions of 3 chat folders (including the folder of interest), with the same exact position order number.

The correct scenario in this case here is it should be removed as this is a unread messages folder, which shouldn't display chats with 0 unread messages. Which means the updateChatPosition is providing the right value but not updateChatLastMessage.

I have uploaded a build of my app that omits use of updateChatLastMessage and so far feedback received from the user, suggests correct behaviour so far.

**Comment by vincentneo** [comment]
Here's the (entire) order of arrival for the same chat id:

1. `updateChatPosition` - main, order 7254849587201271339
2. `updateChatPosition` - folder id 9, order 7254849587201271339
3. `updateChatLastMessage` - empty position array
4. `updateChatPosition` - folder id 8, order 7254849587201271339
5. `updateChatLastMessage` - (folder id 9, order 7257906315425835749), (main, order 7257906315425835749), (folder id 8, order 7257906315425835749)
6. `updateChatPosition` - folder id 8, order 0
7. `updateChatPosition` - folder id 8, order 7257906315425835749
8. `updateChatPosition` - folder id 8, order 0 <- **believed to be where it should have stopped.**
9. `updateChatLastMessage` - (folder id 9, order 7258204317436701476), (main, order 7258204317436701476), (folder id 8, order 7258204317436701476)
10. `updateChatLastMessage` - (folder id 9, order 7258204463465589541), (main, order 7258204463465589541), (folder id 8, order 7258204463465589541)

**Comment by levlam** [comment]
> I have uploaded a build of my app that omits use of updateChatLastMessage and so far feedback received from the user, suggests correct behaviour so far.

No, this is a very incorrect behavior. You must apply all chat positions, received from TDLib. Also, ensure that you completely replace chat positions from `updateChatLastMessage` including removal from missing chat lists.

You need to provide TDLib logs if you are sure that there is a bug in TDLib, but in the mentioned example definitely there were no bug. The chat has received at least two new messages, which are unread, and therefore the chat must be in the Unread folder. After all messages are read in the chat, you will receive `updateChatPosition` with 0 order. `updateChatReadInbox` is an independent update and it can be received before or after that. It is also often delayed to avoid crazy blinking of the unread counter, so you have no way to programmatically check that the update was incorrect by checking `unread_count`.

**Comment by vincentneo** [comment]
> The chat has received at least two new messages

The thing is, the messages were already read on other devices prior to the app being active. If the user reads the message directly on the TDLib-based client, then things work as expected, as I mentioned in the original comment.

> After all messages are read in the chat, you will receive updateChatPosition with 0 order.

Other than the unread_count, the user is able to confirm to me that his unread folder on his phone's Telegram app, has no chats, unlike the TDLib-based client, which contains a bunch of chats, all with 0 unread_count. In fact, when I did the logging, I asked him to replicate the issue by reading all his messages on his phone. 

Regardless, I will check with the user to see if he's willing to contribute a proper TDLib log.

**Comment by levlam** [comment]
You can use `setLogStream` to write the log to a file. For general issues verbosity level 4 is the most suitable.

**Comment by vincentneo** [comment]
Confirmed that issue is on my side.

I would like to apologise for my incorrect accusations on the issue. The entirety of the issue, is a case of a flawed logging system that I have implemented on my end. In fact, I might have been better off without this logging system that I have made, and should have rather get the TDLib log from the user from the start, which would make it clearer to understand.

As always, it is nice to have amazing support from @levlam. Thank you.
